File: basic/src/main/java/guepardoapps/lucahome/basic/controller/ReceiverController.java
package guepardoapps.lucahome.basic.controller;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.IntentFilter;
import android.support.annotation.NonNull;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

import guepardoapps.lucahome.basic.utils.Logger;

public class ReceiverController implements Serializable {
    private static final long serialVersionUID = 2288241732336744506L;

    private static String TAG = ReceiverController.class.getSimpleName();

    private Context _context;
    private List<BroadcastReceiver> _registeredReceiver;

    public ReceiverController(@NonNull Context context) {
        _context = context;
        _registeredReceiver = new ArrayList<>();
    }

    public void RegisterReceiver(
            @NonNull BroadcastReceiver receiver,
            @NonNull String[] actions) {
        IntentFilter downloadStateFilter = new IntentFilter();
        for (String action : actions) {
            downloadStateFilter.addAction(action);
        }

        UnregisterReceiver(receiver);

        _context.registerReceiver(receiver, downloadStateFilter);
        _registeredReceiver.add(receiver);
    }

    public void UnregisterReceiver(@NonNull BroadcastReceiver receiver) {
        for (int index = 0; index < _registeredReceiver.size(); index++) {
            if (_registeredReceiver.get(index) == receiver) {
                try {
                    _context.unregisterReceiver(receiver);
                    _registeredReceiver.remove(index);
                } catch (Exception e) {
                    Logger.getInstance().Error(TAG, e.toString());
                }
                break;
            }
        }
    }

    public void Dispose() {
        for (int index = 0; index < _registeredReceiver.size(); index++) {
            try {
                _context.unregisterReceiver(_registeredReceiver.get(index));
                _registeredReceiver.remove(index);
            } catch (Exception e) {
                Logger.getInstance().Error(TAG, e.toString());
            }
        }
    }
}

File: common/src/main/java/guepardoapps/lucahome/common/controller/ReceiverController.kt
package guepardoapps.lucahome.common.controller

import android.content.BroadcastReceiver
import android.content.Context
import android.content.IntentFilter
import guepardoapps.lucahome.common.utils.Logger

class ReceiverController(private val context: Context) : IReceiverController {
    private val tag: String = ReceiverController::class.java.simpleName

    private val registeredReceiver: ArrayList<BroadcastReceiver> = arrayListOf()

    override fun registerReceiver(registerReceiver: BroadcastReceiver, actions: Array<String>) {
        val intentFilter = IntentFilter()
        for (action in actions) {
            intentFilter.addAction(action)
        }

        unregisterReceiver(registerReceiver)

        context.registerReceiver(registerReceiver, intentFilter)
        registeredReceiver.add(registerReceiver)
    }

    override fun unregisterReceiver(unregisterReceiver: BroadcastReceiver) {
        for (receiver in registeredReceiver) {
            if (receiver == unregisterReceiver) {
                try {
                    context.unregisterReceiver(receiver)
                    registeredReceiver.remove(receiver)
                } catch (exception: Exception) {
                    Logger.instance.error(tag, exception)
                }
                break
            }
        }
    }

    override fun dispose() {
        for (receiver in registeredReceiver) {
            unregisterReceiver(receiver)
        }
    }
}